File: backend/src/utils/distance.js
const { getDistance, getSpeed } = require('geolib');

// Calculate distance between two coordinates in meters
const calculateDistance = (coord1, coord2) => {
  return getDistance(
    { latitude: coord1.lat, longitude: coord1.lng },
    { latitude: coord2.lat, longitude: coord2.lng }
  );
};

// Calculate ETA based on distance and speed
const calculateETA = (distanceMeters, speedKmh) => {
  if (speedKmh === 0) speedKmh = 30; // Default city speed
  const timeHours = (distanceMeters / 1000) / speedKmh;
  return Math.round(timeHours * 60); // Convert to minutes
};

// Check if bus is near a stop (within 100 meters)
const isNearStop = (busLocation, stopLocation, thresholdMeters = 100) => {
  const distance = calculateDistance(busLocation, stopLocation);
  return distance <= thresholdMeters;
};

module.exports = {
  calculateDistance,
  calculateETA,
  isNearStop,
};

File: backend-python/src/utils/distance.py
from geopy.distance import geodesic


def calculate_distance(coord1: dict, coord2: dict) -> float:
    """Return distance in meters between two lat/lng dicts."""
    return geodesic(
        (coord1["lat"], coord1["lng"]),
        (coord2["lat"], coord2["lng"])
    ).meters


def calculate_eta(distance_meters: float, speed_kmh: float) -> int:
    """Return ETA in minutes given distance (m) and speed (km/h)."""
    if speed_kmh <= 0:
        speed_kmh = 30  # default city speed
    time_hours = (distance_meters / 1000) / speed_kmh
    return round(time_hours * 60)


def is_near_stop(bus_location: dict, stop_location: dict, threshold_meters: float = 100) -> bool:
    return calculate_distance(bus_location, stop_location) <= threshold_meters
